xrootd/cmd/xrd-srv/handler.go

```diff
@@ -48,3 +60,153 @@ func (h *handler) Dirlist(sessionID [16]byte, request *dirlist.Request) (xrdprot
+		flag |= os.O_RDWR
+	}
+	if request.Options&xrdfs.OpenOptionsOpenAppend != 0 {
+		flag |= os.O_APPEND
+	}
+	if request.Options&xrdfs.OpenOptionsNew != 0 || request.Options&xrdfs.OpenOptionsDelete != 0 {
+		flag |= os.O_CREATE
+		if request.Options&xrdfs.OpenOptionsDelete == 0 {
+			flag |= os.O_EXCL
+		} else {
+			flag |= os.O_TRUNC
+		}
+	}
+
+	filePath := path.Join(h.basePath, request.Path)
+	if request.Options&xrdfs.OpenOptionsMkPath != 0 {
+		if err := os.MkdirAll(path.Dir(filePath), os.FileMode(request.Mode)); err != nil {
+			return xrdproto.ServerError{
+				Code:    xrdproto.IOError,
+				Message: fmt.Sprintf("An IO error occurred: %v", err),
+			}, xrdproto.Error
+		}
+	}
+
+	file, err := os.OpenFile(filePath, flag, os.FileMode(request.Mode))
+	if err != nil {
+		return xrdproto.ServerError{
+			Code:    xrdproto.IOError,
+			Message: fmt.Sprintf("An IO error occurred: %v", err),
+		}, xrdproto.Error
+	}
+
+	sess, _ := h.sessions.LoadOrStore(sessionID, &session{handles: make(map[xrdfs.FileHandle]*os.File)})
+	s := sess.(*session)
+
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	var handle xrdfs.FileHandle
+
+	for i := 0; i < 255; i++ {
+		handle[0] = byte(i)
+		for j := 0; j < 255; j++ {
+			handle[1] = byte(j)
+			if _, ok := s.handles[handle]; !ok {
+				s.handles[handle] = file
+				// TODO: return stat info if requested.
+				return open.Response{FileHandle: handle}, xrdproto.Ok
+			}
+		}
+	}
+
+	return xrdproto.ServerError{
+		Code:    xrdproto.InvalidRequest,
+		Message: "handle limit exceeded",
+	}, xrdproto.Error
+}
+
+// Close implements server.Handler.Close.
+func (h *handler) Close(sessionID [16]byte, request *xrdclose.Request) (xrdproto.Marshaler, xrdproto.ResponseStatus) {
+	sess, ok := h.sessions.Load(sessionID)
+	if !ok {
+		// This situation can appear if user tries to close without opening any file at all.
+		return xrdproto.ServerError{
+			Code:    xrdproto.InvalidRequest,
+			Message: fmt.Sprintf("Invalid file handle: %v", request.Handle),
+		}, xrdproto.Error
+	}
+	s := sess.(*session)
+	s.mu.Lock()
+	defer s.mu.Unlock()
+	file, ok := s.handles[request.Handle]
+	if !ok {
+		return xrdproto.ServerError{
+			Code:    xrdproto.InvalidRequest,
+			Message: fmt.Sprintf("Invalid file handle: %v", request.Handle),
+		}, xrdproto.Error
+	}
+	delete(s.handles, request.Handle)
+	err := file.Close()
+	if err != nil {
+		return xrdproto.ServerError{
+			Code:    xrdproto.IOError,
+			Message: fmt.Sprintf("An IO error occurred: %v", err),
+		}, xrdproto.Error
+	}
+	return nil, xrdproto.Ok
+}
+
+// Read implements server.Handler.Read.
+func (h *handler) Read(sessionID [16]byte, request *read.Request) (xrdproto.Marshaler, xrdproto.ResponseStatus) {
+	sess, ok := h.sessions.Load(sessionID)
+	if !ok {
+		// This situation can appear if user tries to read without opening any file at all.
+		return xrdproto.ServerError{
+			Code:    xrdproto.InvalidRequest,
+			Message: fmt.Sprintf("Invalid file handle: %v", request.Handle),
```

Comment:
perhaps add more context to this error message, drawing on the comment above?
(or is the comment above only one of the many possible error modes? in that case, yeah, leave it "general/vague" like so.)

Reply:
It's one of the possible routes to the error "invalid file handle":
- user may try to read file with handle that wasn't opened before.
- user may try to read file without opening any file.

In both cases handle is invalid and from user point of view it doesn't matter which case it was.

Reply:
ok. let's leave the "general" string message.